### ISSUE TITLE
[Question] Horizontal scaling?

### ISSUE THREAD

**Issue by RobinJ1995** [F43nd1r/Acrarium#87]
Not a bug report, but just a question (although documenting the answer may be desirable).

I am currently running Acrarium inside Kubernetes, with a MariaDB database. I'd like to ensure Acrarium always has enough capacity to handle incoming requests, but especially that even if one instance goes down, there is always another to take over. For this reason, I would like to have multiple Acrarium pods (connected to the same database) running at all times, with requests load balanced between the two. This in itself is nothing new, but I couldn't find anything about whether Acrarium supports this kind of behaviour.

* Does Acrarium support running several instances that are connected to the same database? (I don't know if Acrarium does any background processing which would  be happening multiple times/conflicting, for example.)
* Does Acrarium have a health check endpoint to ensure it is running correctly, or is my best bet to simply check that the HTTP server returns an HTTP 200 response?

**Comment by F43nd1r** [comment]
Let me start with this: I have no experience with horizontal scaling.
> Does Acrarium support running several instances that are connected to the same database? (I don't know if Acrarium does any background processing which would be happening multiple times/conflicting, for example.)
While never built with this in mind, Acrarium might support it by chance. As far as I know Hibernate uses database transactions, which should prevent most concurrent access problems. 
Only background processing is the automatic bug merging, which is triggered for each report and is transactional. Shouldn't cause issues.
Only issue I can see would be if you use mail notifications. Those would be sent from each instance.

>Does Acrarium have a health check endpoint to ensure it is running correctly, or is my best bet to simply check that the HTTP server returns an HTTP 200 response?
Acra does not include one, but it should be relatively easy to compile spring boot actuator in.

**Comment by RobinJ1995** [comment]
Alright, thank you :)

**Comment by F43nd1r** [comment]
I gave this some more thought: I think running multiple Acrarium instances in parallel would lead to a lot of wasted resources. The part you want to scale up is the [rest backend](https://github.com/F43nd1r/Acrarium/blob/master/acrarium/src/main/kotlin/com/faendir/acra/rest/RestReportInterface.kt), but the bulk of the application is the Vaadin frontend. To be able to scale efficiently, those should be split and scaled independently. I'm not sure if this could be achieved without incresing complexity for monolithic installations.

**Comment by RobinJ1995** [comment]
Yes, in order to be able to scale efficiently, monolithic applications have to broken up into separate microservices. While a single monolithic application is most definitely simpler to develop, it is very hard to scale. I don't personally get enough crash reports to have much need for additional scalability (although it's definitely nice to have), but it also makes it easier to guarantee availability of the service.

It all depends on what the application actually does (I haven't had a proper look at the code), and if it's a lot of effort and you're the only major contributor on this project it may simply not be worth your time, but in general monolithic application architectures don't tend to stand the test of time very well.

So without looking at the source code, these are the assumptions I'm making based on what I know Acrarium does;
* Vaadin frontend + API (monolith)
  * Generates HTML/Javascript (Vaadin)
  * Performs database transactions
  * Performs (background?) tasks (e.g. sending e-mail)
* Database

How you could potentially split this up is;
* Vaadin frontend
  * Serve static content
* REST API:
  * Perform database transactions
* Task runner
  * Performs background tasks
  * To make the task runner scalable, either a database table or something like Redis could be used to obtain locks and ensure a task only gets run once (alternatively, you could go for a message queue based approach)
* Database

In this architecture, the REST API would only perform actions that happen as result of a request. Since any single request is only ever routed to a single instance, there are no concurrency issues here.
Background tasks could be run by the task runner, which would need some sort of locking mechanism (database table, Redis, ...) to ensure that if there are multiple instances of the service running, any task only gets executed once.
If the REST API needs to be able to trigger background tasks, a message queue (RabbitMQ, Apache Pulsar, Redis pub/sub, ... or again, if simplicity is desired a database table can be (ab)used for this purpose, too) can be used for this purpose, from which the task runner can then read messages and execute tasks.

**Comment by F43nd1r** [comment]
> if it's a lot of effort and you're the only major contributor on this project it may simply not be worth your time

Yes, this redesign is definetly not going to happen in the near future.

I personally think the topic is interesting, so I might try this sometime, but certainly not before 1.0 stable release.

For now, some arguments why I think acrarium doesn't really need horizontal scaling:
 - Spring Boot works well with vertical scaling, parallel request processing is solid. You'll need _a lot_ of crashes to make this not a valid solution
 - ACRA is built with low (network) availability in mind, so 100% uptime isn't required.
